### PR TITLE
fix: properly update branch version

### PR DESCRIPTION
**Pull request by alemidev**
basically we used to merge and update local version while generating deltas, but it shouldnt happen (we added `ack` for that)

this changes how `hash` works however: it now reflects the state of the buffer _before_ applying this change, not after. i think its a bit less useful but i dont see other ways

### CODE REVIEW

**Comment by ghyatzo** [comment]
I will have to think this through.

If I recall correctly, we used to incrementally add changes, since we are grabbing from and iterator of changes. It might be more than one.

So the `step_ver` was the version after one step, and it might not be at the frontier. We keep this so that the next time we can get the next change in the sequence correctly.

I will have to go back in time a bit and remember how it was before, and how it changed when `ack` was added to figure out what went wrong though.

I'd like to err on the side of caution and not merge this until we are (maybe it's just me?) more sure this is the actual fix.

**Comment by zaaarf** [comment]
I don't think it matters all that much whether `hash` represents before or after. I mean, sure, obviously it requires changes in the plugin impl, but it's not like we set `hash` every time anyway. Instead of being caught in the current iteration, a desync will be caught in the one after that. I don't think it necessarily makes `hash` less useful.

The code/logic looks fine to me, but I'll leave it to @ghyatzo's judgement to assess whether it's the change that fixes it since he's the resident CRDT connoisseur.

Just one thing: why'd you change `Default::default()`to specify the impl name? Is this going to be consistent throughout the codebase now?

**Review by zaaarf — APPROVED** [review]

**Review by ghyatzo — APPROVED** [review]
I'm as convinced as I can by just reading code. Looks good, guess we need to test it now.

**Pull request merged**
